ultralytics/utils/ops.py

```diff
@@ -263,6 +265,7 @@ def non_max_suppression(
         # Detections matrix nx6 (xyxy, conf, cls)
         box, cls, mask = x.split((4, nc, nm), 1)
 
+        proba = x[..., 4:] if soft_label else None  # all class probabilities
```

Comment:
Probably should be moved further down since `x` is still getting updated after this line.

Reply:
Actually that's why this is here. L247 will truncate the probabilities to only the max value, so after the next conditional, the values for all class probabilities are not longer available in `x`

Reply:
I guess you have to filter `proba` using `i` after L274 then, because the lines prior is filtering `x` and the correspondence might be lost between the two.

Reply:
hm, I do us `i` on L310 with `proba` but wondering if it could be the `argsort` from L285 that might cause issues. Guessing you're referring to the problem shared in the comment from O-J1? Originally I was _only_ using this for `predict` but it would be cool to see it working for `val` as well, so open to any ideas/suggestions.

Reply:
Yes. 

What I mean is that the following lines are filtering `x`:
https://github.com/ultralytics/ultralytics/blob/68dbf90bf88f8ee03c6fef3bd5bee0454b4dbf15/ultralytics/utils/ops.py#L269-L275

which means the `x` that corresponds to `proba` is lost. `proba` has to go through the same exact filter to maintain the correspondence. I was referring to `i` in 270. So there should be the `proba = proba[i]` after L271 to handle the first case, and `proba = proba[conf.view(-1) > conf_thres]` after L274 to handle the second case.

Reply:
Similarly for L278, L285.